piptools/resolver.py

```diff
@@ -25,27 +30,31 @@ class RequirementSummary:
     Summary of a requirement's properties for comparison purposes.
     """
 
-    def __init__(self, ireq):
+    def __init__(self, ireq: InstallRequirement) -> None:
         self.req = ireq.req
         self.key = key_from_ireq(ireq)
         self.extras = frozenset(ireq.extras)
         self.specifier = ireq.specifier
 
-    def __eq__(self, other):
-        return (
-            self.key == other.key
-            and self.specifier == other.specifier
-            and self.extras == other.extras
-        )
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, self.__class__):
+            return (
+                self.key == other.key
+                and self.specifier == other.specifier
+                and self.extras == other.extras
+            )
+        return NotImplemented
```

Comment:
Also, I believe it's better to have an inversed guard-expression as this reduces the number of nested lines
```suggestion
        if not isinstance(other, self.__class__):
            return NotImplemented

        return (
            self.key == other.key
            and self.specifier == other.specifier
            and self.extras == other.extras
        )
```

Reply:
Yeah, looks better now. Thanks!